### PR TITLE
fix(vpc): add resolver for reverse dns

### PR DESCRIPTION
**Pull request by yfodil**

### CODE REVIEW

File: scaleway/helpers_instance.go [hunk 1]
```diff
@@ -603,17 +604,31 @@ func isInstanceIPReverseResolved(ctx context.Context, instanceAPI *instance.API,
...
+
+	for {
+		select {
+		case <-ticker:
+			address, err := r.LookupHost(ctx, reverse)
+			if err != nil {
+				if ctx.Err() == context.DeadlineExceeded {
+					return false
+				}
+			} else if slices.Contains(address, res.IP.Address.String()) {
```
Comment (remyleone) [inline scaleway/helpers_instance.go:627]: else return slices.Contains(address, res.IP.Address.String())

File: scaleway/helpers_vpcgw.go [hunk 2]
```diff
@@ -110,19 +111,33 @@ func isGatewayIPReverseResolved(ctx context.Context, api *vpcgw.API, reverse str
...
+			return conn, err
+		},
 	}
-	for range ticker {
-		address, err := r.LookupHost(ctx, reverse)
-		if err != nil {
-			if ctx.Err() == context.DeadlineExceeded && IP.Address.String() == address[0] {
-				return false
+
+	for {
```
Comment (remyleone) [inline scaleway/helpers_vpcgw.go:126]: Could this function be extracted? It is exactly the same logic in both cases?

File: scaleway/resource_instance_ip_reverse_dns.go [hunk 3]
```diff
@@ -64,11 +64,14 @@ func resourceScalewayInstanceIPReverseDNSCreate(ctx context.Context, d *schema.R
 			IP:   res.IP.ID,
 		}
 
-		reverse := d.Get("reverse").(string)
-		if reverse == "" {
-			updateReverseReq.Reverse = &instance.NullableStringValue{Null: true}
+		if reverse, ok := d.GetOk("reverse"); ok {
+			if isInstanceIPReverseResolved(ctx, instanceAPI, reverse.(string), defaultInstanceIPReverseDNSTimeout, res.IP.ID, zone) {
```
Comment (Codelax) [inline scaleway/resource_instance_ip_reverse_dns.go:68]:
```suggestion
			if isInstanceIPReverseResolved(ctx, instanceAPI, reverse.(string), d.Timeout(schema.TimeoutCreate), res.IP.ID, zone) {
```

Comment (Codelax) [reply scaleway/resource_instance_ip_reverse_dns.go:68]: TimeoutCreate should also be added to the schema

File: scaleway/resource_instance_ip_reverse_dns.go [hunk 4]
```diff
@@ -117,11 +120,14 @@ func resourceScalewayInstanceIPReverseDNSUpdate(ctx context.Context, d *schema.R
 			IP:   ID,
 		}
 
-		reverse := d.Get("reverse").(string)
-		if reverse == "" {
-			updateReverseReq.Reverse = &instance.NullableStringValue{Null: true}
+		if reverse, ok := d.GetOk("reverse"); ok {
+			if isInstanceIPReverseResolved(ctx, instanceAPI, reverse.(string), defaultInstanceIPReverseDNSTimeout, ID, zone) {
```
Comment (Codelax) [inline scaleway/resource_instance_ip_reverse_dns.go:124]:
```suggestion
			if isInstanceIPReverseResolved(ctx, instanceAPI, reverse.(string), d.Timeout(schema.TimeoutUpdate), ID, zone) {
```

File: scaleway/resource_vpc_public_gateway_ip_reverse.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		tflog.Debug(ctx, fmt.Sprintf("updating IP %q reverse to %q\n", d.Id(), d.Get("reverse")))
+
+		updateReverseReq := &vpcgw.UpdateIPRequest{
+			Zone: zone,
+			IPID: res.ID,
+		}
+
+		reverse := d.Get("reverse").(string)
+		if len(reverse) > 0 {
+			if isGatewayIPReverseResolved(ctx, vpcgwAPI, reverse, defaultVPCPublicGatewayIPReverseDNSTimeout, res.ID, zone) {
```
Comment (Codelax) [inline scaleway/resource_vpc_public_gateway_ip_reverse.go:68]:
```suggestion
			if isGatewayIPReverseResolved(ctx, vpcgwAPI, reverse, d.Timeout(schema.TimeoutCreate), res.ID, zone) {
```

File: scaleway/resource_vpc_public_gateway_ip_reverse.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		tflog.Debug(ctx, fmt.Sprintf("updating IP %q reverse to %q\n", d.Id(), d.Get("reverse")))
+
+		updateReverseReq := &vpcgw.UpdateIPRequest{
+			Zone: zone,
+			IPID: ID,
+		}
+
+		reverse := d.Get("reverse").(string)
+		if len(reverse) > 0 {
+			if isGatewayIPReverseResolved(ctx, vpcgwAPI, reverse, defaultVPCPublicGatewayIPReverseDNSTimeout, ID, zone) {
```
Comment (Codelax) [inline scaleway/resource_vpc_public_gateway_ip_reverse.go:123]:
```suggestion
			if isGatewayIPReverseResolved(ctx, vpcgwAPI, reverse, d.Timeout(schema.TimeoutUpdate), ID, zone) {
```

File: scaleway/resource_vpc_public_gateway_ip_reverse.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	return &schema.Resource{
+		CreateContext: resourceScalewayVPCPublicGatewayIPReverseDNSCreate,
+		ReadContext:   resourceScalewayVPCPublicGatewayIPReverseDNSRead,
+		UpdateContext: resourceScalewayVPCPublicGatewayIPReverseDNSUpdate,
+		DeleteContext: resourceScalewayVPCPublicGatewayIPReverseDNSDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Default: schema.DefaultTimeout(defaultVPCPublicGatewayIPReverseDNSTimeout),
```
Comment (Codelax) [inline scaleway/resource_vpc_public_gateway_ip_reverse.go:24]:
```suggestion
			Default: schema.DefaultTimeout(defaultVPCPublicGatewayIPReverseDNSTimeout),
			Create:  schema.DefaultTimeout(defaultVPCPublicGatewayIPReverseDNSTimeout),
			Update:  schema.DefaultTimeout(defaultVPCPublicGatewayIPReverseDNSTimeout),
```

**Review by Codelax — CHANGES_REQUESTED** [review]

**Review by Codelax — APPROVED** [review]

**Pull request merged**